Tasks/AzureFileCopyV2/AzureFileCopy.ps1

```diff
@@ -0,0 +1,250 @@
+    # /Z: Journal file location
+    # /V: AzCopy verbose logs file location
+
+    Write-Verbose "Using default AzCopy arguments for uploading to blob storage"
+
+    $logFileName = "AzCopyVerbose_" + [guid]::NewGuid() + ".log"
+    $logFilePath = Join-Path -Path $azCopyLocation -ChildPath $logFileName
+
+    $additionalArgumentsForBlobCopy = "/XO /Y /SetContentType /Z:`"$azCopyLocation`" /V:`"$logFilePath`""
+
+    # Add more arguments if required
+
+    # Premium storage accounts only support page blobs
+    if($isPremiumStorage)
+    {
+        Write-Verbose "Setting BlobType to page for Premium Storage account."
+        $additionalArgumentsForBlobCopy += " /BlobType:page"
+    }
+
+    # $root container does not support sub folders. So excluding recursive copy option for $root container.
+    if($containerName -ne '$root')
+    {
+        Write-Verbose "Adding argument for recursive copy"
+        $additionalArgumentsForBlobCopy += " /S"
+    }
+}
+
+Check-ContainerNameAndArgs -containerName $containerName -additionalArguments $additionalArgumentsForBlobCopy
+
+# Uploading files to container
+Upload-FilesToAzureContainer -sourcePath $sourcePath `
+                             -storageAccountName $storageAccount `
+                             -containerName $containerName `
+                             -blobPrefix $blobPrefix `
+                             -blobStorageEndpoint $blobStorageEndpoint `
+                             -storageKey $storageKey `
+                             -azCopyLocation $azCopyLocation `
+                             -additionalArguments $additionalArgumentsForBlobCopy `
+                             -destinationType $destination `
+                             -useDefaultArguments $useDefaultArgumentsForBlobCopy `
+                             -azCopyLogFilePath $logFilePath
+
+# Complete the task if destination is azure blob
+if ($destination -eq "AzureBlob")
+{
+    # Get URI and SaSToken for output if needed
+    if(-not [string]::IsNullOrEmpty($outputStorageURI))
+    {
+        $storageAccountContainerURI = $storageContext.BlobEndPoint + $containerName
+        Write-Host "##vso[task.setvariable variable=$outputStorageURI;]$storageAccountContainerURI"
+    }
+    if(-not [string]::IsNullOrEmpty($outputStorageContainerSASToken))
+    {
+        $storageContainerSaSToken = New-AzureStorageContainerSASToken -Container $containerName -Context $storageContext -Permission r -ExpiryTime (Get-Date).AddHours($defaultSasTokenTimeOutInHours)
+        Write-Host "##vso[task.setvariable variable=$outputStorageContainerSASToken;]$storageContainerSasToken"
+    }
+    Write-Verbose "Completed Azure File Copy Task for Azure Blob Destination"
+    return
+}
+
+# Copying files to Azure VMs
+try
+{
+    # Normalize admin username
+    if($vmsAdminUserName -and (-not $vmsAdminUserName.StartsWith(".\")) -and ($vmsAdminUserName.IndexOf("\") -eq -1) -and ($vmsAdminUserName.IndexOf("@") -eq -1))
```

Comment:
check if $vmsAdminUserName.trim required

Reply:
I haven't tried creating a VM like this, but azure allows spaces in username and password of vm's. So I think we should keep it same